### PR TITLE
Patrick_NotificationOfUser

### PR DESCRIPTION
**Pull request by khoainx**
@minhquan4080 please review and give me your opinion!

### CODE REVIEW

File: app/controllers/user/registrations_controller.rb [hunk 1]
```diff
@@ -70,8 +70,39 @@ def update_password
...
+    @noti = Notification.get_noti
+    if last_id.nil?
+      @noti = @noti.limit(5)
+    else
+      if (last_id+5) > @noti.count
+        @noti = @noti[last_id..@noti.count]
+      else
+        @noti = @noti[last_id..last_id+4]
+      end
+    end
```
Comment (minhquan4080) [inline app/controllers/user/registrations_controller.rb]: @noti = ((last_id + 5) > @noti.count) ? @noti[last_id..@noti.count] : @noti[last_id..last_id+4]

File: app/views/user/registrations/_notification.html.slim [hunk 2]
```diff
@@ -0,0 +1,14 @@
+- noti.each do |noti|
```
Comment (minhquan4080) [inline app/views/user/registrations/_notification.html.slim]: @noti.each do | noti |

**Comment by minhquan4080** [comment]
@khoainx  please check my comment

File: app/controllers/api/notification_controller.rb [hunk 3]
```diff
@@ -1,11 +1,11 @@
 class Api::NotificationController < ApplicationController
   def index
-     @notification =  Notification.where(author: current_user.id).order(id: 'desc')
+     @notification =  Notification.get_noti
```
Comment (lhhieu89) [inline app/controllers/api/notification_controller.rb]: Please define the function name clearly, like `Notification._by_user(current_user)` or `Notification.by_user_id(current_user.id)`

File: app/controllers/application_controller.rb [hunk 4]
```diff
@@ -3,7 +3,11 @@ class ApplicationController < ActionController::Base
   # For APIs, you may want to use :null_session instead.
   protect_from_forgery with: :exception
   after_filter :session_location
+  before_filter :set_current_user
 
+  def set_current_user
+    Notification.current_user = current_user
+  end
```
Comment (lhhieu89) [inline app/controllers/application_controller.rb]: You don't need to define this function here, you can get `current_user` from the controller and send it to the model,...

File: app/controllers/user/registrations_controller.rb [hunk 5]
```diff
@@ -70,8 +70,35 @@ def update_password
...
     end
   end
-
+  def loadmorenoti
+    last_id = params[:last_id].to_i
+    @noties = Notification.get_noti
+    if last_id.nil?
+      @noties = @notie.limit(5)
+    else
+      @noties = ((last_id + 5) < @noties.count) ? @noties[last_id..@noti.count] : @noties[last_id..last_id+4]
```
Comment (lhhieu89) [inline app/controllers/user/registrations_controller.rb]: Please use pagination to get more data, we are using `kaminari`

File: app/models/notification.rb [hunk 6]
```diff
@@ -1,7 +1,17 @@
...
   enum notification_type: {commented: 1, liked: 2, rated: 3}
+  def self.get_noti
+    Notification.where(author: current_user.id).order(id: 'desc')
+  end
+  def self.get_unread
+    Notification.where(author: current_user.id, seen: false).order(id: 'desc')
+  end
+  def self.seen
+    Notification.where(author: current_user.id).update_all(seen: true)
+  end
```
Comment (lhhieu89) [inline app/models/notification.rb]: Please try to use `scope`

File: app/views/user/registrations/_loadmorenoti.html.slim [hunk 7]
```diff
@@ -0,0 +1,14 @@
...
+      - if noti.commented?
+        i class="fa fa-comments-o "
+      - elsif noti.liked?
+        i class="fa fa-thumbs-o-up"
+      - else
+        i class="fa fa-star"
+      h5 #{noti.user.username} #{noti.message.humanize}
+    // = image_tag ''
+    .noti-info-right
+      = image_tag ("#{noti.user.photo}"), title: "#{noti.user.username.capitalize}: #{noti.user.email} "
```
Comment (lhhieu89) [inline app/views/user/registrations/_loadmorenoti.html.slim]: Just `noti.user.photo`, don't need to use `"#{noti.user.photo}"`

File: app/controllers/user/registrations_controller.rb [hunk 8]
```diff
@@ -70,8 +70,35 @@ def update_password
       render "edit_password"
     end
   end
-
+  def loadmorenoti
+    last_id = params[:last_id].to_i
+    @noties = Notification.get_noti
+    if last_id.nil?
+      @noties = @notie.limit(5)
```
Comment (lhhieu89) [inline app/controllers/user/registrations_controller.rb]: This is `@noties.limit(5)` not `@notie.limit(5)`

**Comment by khoainx** [comment]
@lhhieu89 please review again and give me your opinion!

File: app/assets/javascripts/notification.js.erb [hunk 9]
```diff
@@ -1,3 +1,26 @@
+function load_more_notification(pageno){
+  var last_id = $('.notification-element').last().index();
+  $.ajax({
+    url: "/user/loadmorenotification",
+    data: {pageno: pageno},
+    success: function(result){
+      console.log(result);
```
Comment (minhquan4080) [inline app/assets/javascripts/notification.js.erb]: remove console.log

File: app/assets/javascripts/notification.js.erb [hunk 10]
```diff
@@ -1,3 +1,26 @@
...
+  })
+}
+$(document).ready(function() {
+  var currPage = 1;
+  $("a.next").click(function() {
+    load_more_notification(++currPage);
+  });
+  $(window).scroll(function() {
+    if($(window).scrollTop() + $(window).height() > $(document).height() - 200){
+      console.log('fdfdf');
```
Comment (minhquan4080) [inline app/assets/javascripts/notification.js.erb]: remove console.log

File: app/models/notification.rb [hunk 11]
```diff
@@ -3,5 +3,17 @@ class Notification < ActiveRecord::Base
   belongs_to :idea
 
   enum notification_type: {commented: 1, liked: 2, rated: 3}
+  scope :by_user_id, lambda{|id| where(author: id).order(id: 'desc')}
+  scope :user_unread_notifications, lambda{|id| where(author: id, seen: false).order(id: 'desc')}
+  scope :seen_all_user_notifications, lambda{|id| where(author: id).update_all(seen: true)}
+  # def self.notifications_of_user
```
Comment (minhquan4080) [inline app/models/notification.rb]: If don't use , please remove it.

**Comment by khoainx** [comment]
@minhquan4080  I fixed it and move to new branch: patrick_loadmorepopup

File: app/controllers/user/registrations_controller.rb [hunk 12]
```diff
@@ -70,8 +70,37 @@ def update_password
...
-
+  def loadmorenotification
+    # last_id = params[:last_id].to_i
+    pageno = params[:pageno].to_i
+    @notifications = Notification.by_user_id(current_user.id).page(params[:page])
+    if pageno.nil?
+      @notifications = @notifications.page(1)
+    else
+      @notifications = @notifications.page(pageno)
+    end
```
Comment (lhhieu89) [inline app/controllers/user/registrations_controller.rb]: I think you don't need 5 lines above, just use `params[:page]` instead of `params[:pageno]`

**Pull request closed without merging**